src/fetcher.js

```diff
@@ -1110,8 +1113,42 @@ class Fetcher {
     options.method = method
     options.body = options.data || options.body
     options.force = true
+    const fetcher = this
 
-    return this.fetch(uri, options)
+    if (options.body && !options.contentType) {
+      throw new Error('Web operation sending data must have a defined contentType.')
+    }
+    if (options.contentType) {
+      options.headers = options.headers || {}
+      options.headers['content-type'] = options.contentType
+    }
+    if (Fetcher.withCredentials(uri, options)) {
+      options.credentials = 'include' // Otherwise coookies are not sent
+    }
+
+    return new Promise(function (resolve, reject) {
+      fetcher._fetch(uri, options).then(response => {
+        if (response.ok) {
+          if (response.body) {
+            response.text().then(data => {
+              response.responseText = data
+              resolve(response)
+            })
+          } else {
+            resolve(response)
+          }
+        } else {
+          let msg = 'Web error: ' + response.status
+          if (response.statusText) msg += ' (' + response.statusText + ')'
+          msg += ' on ' + method + ' of <' + uri + '>'
+          if (response.responseText) msg += ': ' + response.responseText
+          reject(new Error(msg))
```

Comment:
Extract to a class?

```
class FetcherError extends Error {
  constructor(uri, options, response) {
    super();
    this.uri = uri;
    this.options = options;
    this.response = response;
    this.message = this.toString();

    // Optionally with improved console logging (https://stackoverflow.com/a/32749533/1630540)
    this.name = this.constructor.name;
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, this.constructor);
    } else { 
      this.stack = (new Error(message)).stack; 
    }
  }

  toString() {
    let msg = 'Web error: ' + this.response.status
    if (this.response.statusText) msg += ' (' + this.response.statusText + ')'
      msg += ' on ' + method + ' of <' + uri + '>'
    if (this.response.responseText) msg += ': ' + this.response.responseText
    return msg
  }
}
```

Reply:
Good idea.. Maybe HTTPError subclass of FetcherError (WebError?) subclass of Error.
I'd like to land this branch, though now, and do that in a separate branch.    And include where there is one an HTTP error body, parsed if it is turtle, and including the nested error objects where they exist.  (Is there a convention for this/  error.error.erro... ? error.cause ?)

Reply:
HTTPError < FetcherError seems appropriate.

The property `original` comes to mind since its usage in event targets. Calling it something along the lines of `error` or `serverError` begs the question whether the response body of a failed request actually is an error object or some other resource. So perhaps using the `body` property like the Response class is better.

As an example use-case around body parsing; Our API currently returns custom error resources via nquads, with their subject set to the resource in question, so the client renders an error view. When the request succeeds at a later point, the data is overwritten so the actual resource is rendered. The design enables the client to render errors without having to implement logic to retrieve the status of past requests and extrapolate whether an error has occurred.

Maybe something like the following pattern; If the response Content type is a valid RDF format, parse the body. If the body contains a document-relative resource (ttl `_:` IIRC), add that resource (statement array or converted to js?) to the error object under some key e.g. `resource`. Process additional statements included in the response like the request succeeded (Flagged opt in/out).

On a slightly less related note, I came across [RFC 7807](https://tools.ietf.org/html/rfc7807) which defines a JSON and XML objects for "HTTP API problems" which could act as a class model reference (though SHACL validation reports might also be a common body type).